AsyncDisplayKit/Details/ASRangeController.mm

```diff
@@ -159,42 +82,67 @@ - (void)updateVisibleNodeIndexPaths
     return;
   }
 
-  NSArray *indexPaths = [_delegate rangeControllerVisibleNodeIndexPaths:self];
+  NSArray *visibleNodePaths = [_delegate rangeControllerVisibleNodeIndexPaths:self];
+  NSSet *visibleNodePathsSet = [NSSet setWithArray:visibleNodePaths];
   CGSize viewportSize = [_delegate rangeControllerViewportSize:self];
 
-  if ([_layoutController shouldUpdateWorkingRangesForVisibleIndexPath:indexPaths viewportSize:viewportSize]) {
-    [_layoutController setVisibleNodeIndexPaths:indexPaths];
-    NSSet *workingRangeIndexPaths = [_layoutController workingRangeIndexPathsForScrolling:_scrollDirection viewportSize:viewportSize];
-    NSSet *visibleRangeIndexPaths = [NSSet setWithArray:indexPaths];
-
-    NSMutableSet *removedIndexPaths = _workingRangeIsValid ? [_workingRangeIndexPaths mutableCopy] : [NSMutableSet set];
-    [removedIndexPaths minusSet:workingRangeIndexPaths];
-    [removedIndexPaths minusSet:visibleRangeIndexPaths];
-    if (removedIndexPaths.count) {
-      NSArray *removedNodes = [_delegate rangeController:self nodesAtIndexPaths:[removedIndexPaths allObjects]];
-      [removedNodes enumerateObjectsUsingBlock:^(ASCellNode *node, NSUInteger idx, BOOL *stop) {
-        [self removeNodeFromWorkingRange:node];
-      }];
-    }
-
-    NSMutableSet *addedIndexPaths = [workingRangeIndexPaths mutableCopy];
-    [addedIndexPaths minusSet:_workingRangeIndexPaths];
-    [addedIndexPaths minusSet:visibleRangeIndexPaths];
-    if (addedIndexPaths.count) {
-      NSArray *addedNodes = [_delegate rangeController:self nodesAtIndexPaths:[addedIndexPaths allObjects]];
-      [addedNodes enumerateObjectsUsingBlock:^(ASCellNode *node, NSUInteger idx, BOOL *stop) {
-        [self addNodeToWorkingRange:node];
-      }];
+  // the layout controller needs to know what the current visible indices are to calculate range offsets
+  [_layoutController setVisibleNodeIndexPaths:visibleNodePaths];
+
+  for (NSInteger i = 0; i < ASLayoutRangeTypeCount; i++) {
+    ASLayoutRangeType rangeType = (ASLayoutRangeType)i;
+    id rangeKey = @(rangeType);
+
+    // this delegate decide what happens when a node is added or removed from a range
+    id<ASRangeDelegate> rangeDelegate = _rangeTypeHandlers[rangeKey];
```

Comment:
nit, it may be better to wrap below code into a function and call with different parameters.